// preact-cli/lib/commands/build.js
'use strict';

exports.__esModule = true;

var _path = require('path');

var _es6Promisify = require('es6-promisify');

var _es6Promisify2 = _interopRequireDefault(_es6Promisify);

var _rimraf = require('rimraf');

var _rimraf2 = _interopRequireDefault(_rimraf);

var _asyncCommand = require('../lib/async-command');

var _asyncCommand2 = _interopRequireDefault(_asyncCommand);

var _webpackConfig = require('../lib/webpack-config');

var _webpackConfig2 = _interopRequireDefault(_webpackConfig);

var _runWebpack = require('../lib/run-webpack');

var _runWebpack2 = _interopRequireDefault(_runWebpack);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

exports.default = (0, _asyncCommand2.default)({
	command: 'build [src] [dest]',

	desc: 'Create a production build in build/',

	builder: {
		src: {
			description: 'Entry file (index.js)',
			default: 'src'
		},
		dest: {
			description: 'Directory root for output',
			default: 'build'
		},
		production: {
			description: 'Create a minified production build.',
			alias: 'p',
			default: true
		},
		prerender: {
			description: 'Pre-render static app content.',
			default: true
		},
		clean: {
			description: 'Clear output directory before building.',
			default: true
		},
		json: {
			description: 'Generate build statistics for analysis.',
			default: false
		}
	},

	handler(argv) {
		return _asyncToGenerator(function* () {
			var config = (0, _webpackConfig2.default)(argv);

			if (argv.clean) {
				var dest = (0, _path.resolve)(argv.cwd || process.cwd(), argv.dest || 'build');
				yield (0, _es6Promisify2.default)(_rimraf2.default)(dest);
			}

			var stats = yield (0, _runWebpack2.default)(false, config);
			(0, _runWebpack.showStats)(stats);

			if (argv.json) {
				yield (0, _runWebpack.writeJsonStats)(stats);
			}
		})();
	}
});
